### PR TITLE
Fix admin datetime fields to respect store timezone

### PR DESCRIPTION
**Pull request by 55728**
Fixes #13393

## Problem

The `datetime-local` input fields in the admin panel do not include timezone information. The browser displays and accepts input in the user's local timezone, but the server stores the value as-is without timezone conversion. This causes promotions to start/expire and products to become available/discontinue at incorrect times when the store's timezone differs from UTC.

## Solution

Use the existing `current_store.preferred_timezone` to interpret the submitted datetime values, consistent with how order date filters and other datetime inputs are already handled in the admin panel (see `OrdersFiltersHelper`, `ResourceController`, `AnalyticsConcern`).

- **On save**: Parse datetime fields in the store's timezone before persisting
- **On display**: Convert stored UTC values to the store's timezone for form fields, with timezone name shown in help text
- **Shared helper**: `parse_datetime_in_store_timezone` extracted to `BaseController` for reuse across controllers
- **I18n**: Timezone help text uses `Spree.t` key

### Why store timezone instead of browser timezone?

The issue suggests capturing the user's browser timezone via a hidden field. However, Spree already has a per-store timezone preference (`Spree::Store#preferred_timezone`) used throughout the admin (dashboard analytics, order filters, reports, CSV exports). Using the store timezone ensures all admins see consistent scheduling regardless of which device or location they use.

### Affected fields
- **Promotions**: `starts_at`, `expires_at`
- **Products**: `available_on`, `discontinue_on`, `make_active_at`

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: spree/admin/app/views/spree/admin/promotions/form/_settings.html.erb [hunk 1]
```diff
@@ -2,6 +2,17 @@
...
 <% end %>
 
-<%= f.spree_datetime_field :starts_at %>
+<%# datetime-local inputs have no timezone; render values in the store's timezone so admins see and submit values in the store's local time. %>
+<% store_timezone = current_store.preferred_timezone %>
+<%
+  to_store_local   = ->(value) { value&.in_time_zone(store_timezone)&.strftime('%Y-%m-%dT%H:%M') }
+  starts_at_local  = to_store_local.call(@promotion.starts_at)
+  expires_at_local = to_store_local.call(@promotion.expires_at)
+  tz_help          = Spree.t('admin.datetime_field_timezone_help', zone: store_timezone)
```
[the remark this answers is not in the document]
Comment (55728) [reply spree/admin/app/views/spree/admin/promotions/form/_settings.html.erb]: Thanks for the suggestion! With only two views using this pattern, I think the inline lambda keeps things simple and readable. If more datetime-local fields are added in the future, extracting to a helper would make sense.

File: spree/admin/app/controllers/spree/admin/base_controller.rb [hunk 2]
```diff
@@ -88,6 +88,23 @@ def current_timezone
...
+
+        fields.each do |field|
+          value = attrs[field]
+          next if value.blank?
+
+          attrs[field] = begin
+            zone.parse(value.to_s)
+          rescue ArgumentError
+            value
+          end
```
[the remark this answers is not in the document]
Comment (55728) [reply spree/admin/app/controllers/spree/admin/base_controller.rb]: Thanks for the catch! You're right — `ActiveSupport::TimeZone#parse` returns `nil` for unparseable values without raising, which would have silently cleared the field. Applied your suggested fix in 63822ee041: we now only assign back when `parsed_value.present?`, leaving the original value untouched (so normal validation / type-casting handles it) when parsing fails or returns nil.

File: spree/admin/app/views/spree/admin/promotions/form/_settings.html.erb (hunk 1 above)
[the remark this answers is not in the document]
Comment (55728) [reply spree/admin/app/views/spree/admin/promotions/form/_settings.html.erb]: Good catch! Addressed in 63822ee — both promotions and products form partials now resolve the zone via `ActiveSupport::TimeZone[current_store.preferred_timezone] || Time.zone`, and `tz_help` uses the resolved `zone.name`. This keeps the view side symmetric with `BaseController#parse_datetime_in_store_timezone`.

**Review by damianlegawiec — COMMENTED** [review]
## Review Summary

This PR fixes a real and well-understood problem: `datetime-local` HTML inputs strip timezone information, so naive datetime strings get persisted as UTC instead of being interpreted in the store's timezone. The approach — using `current_store.preferred_timezone` — is consistent with how timezone conversion is already done throughout the admin (OrdersFiltersHelper, ResourceController date range filters, AnalyticsConcern, DashboardController analytics grouping).

**CI:** All 28 checks pass (PostgreSQL, MySQL, Brakeman, spelling, CLA).

### What works well

- **Correct architectural decision.** Using the store timezone rather than the browser timezone ensures all admins see consistent scheduling regardless of their physical location. This matches the existing Spree pattern.
- **Clean separation of concerns.** Parse on save (`parse_datetime_in_store_timezone` in BaseController), convert on display (view partials). The helper is well-placed for reuse by other controllers.
- **Nil-safe parsing.** Only overwrites `attrs[field]` when `parsed_value.present?`, so unparseable or blank values fall through to normal Rails validation/type-casting. Good defensive coding (addressed from Copilot review feedback).
- **Timezone fallback.** Both the controller helper and view partials fall back to `Time.zone` when the store's `preferred_timezone` is blank or invalid.
- **Round-trip correctness.** A stored UTC value → displayed as store-local → resubmitted unchanged → parsed back to the same UTC instant. The conversion is symmetric.
- **Good test coverage.** Tests use `Asia/Tokyo` (UTC+9) which makes the offset math easy to verify. Both the "stores as UTC" assertion and the blank-value edge case are covered.

### Suggestions (non-blocking)

1. **View-level duplication.** The timezone conversion lambda + `tz_help` setup block is copy-pasted between `_status.html.erb` and `_settings.html.erb`. With only two call sites this is tolerable, but if a third appears, extracting a helper like `admin_datetime_in_store_tz(value)` into an admin helper module would reduce drift risk. The `spree_datetime_field` form builder method itself could also be enhanced to accept a `timezone:` option that handles the conversion internally — that would be the most DRY long-term solution.

2. **`min`/`max` attribute format.** The PR correctly converts `min:` and `max:` from raw `TimeWithZone` objects to formatted store-local strings (`discontinue_on_local`, `make_active_at_local`). This is necessary because `datetime-local` inputs expect `YYYY-MM-DDTHH:MM` format for `min`/`max` constraints. Worth noting in a code comment for future maintainers, since the original code passed raw model values that happened to work by accident (Rails' `to_s` on Time objects).

3. **Seconds truncation.** The `strftime('%Y-%m-%dT%H:%M')` format drops seconds. This is fine for `datetime-local` inputs (which typically have minute granularity), but if any of these fields ever store meaningful seconds, they'd be silently truncated on the next edit. Low risk given the domain (promotion scheduling, product availability), but worth being aware of.

4. **Promotion `expires_at` help text concatenation.** The line `help: "Leave this field blank for no expiration. #{tz_help}"` mixes a hardcoded English string with an I18n key. The original code already had the hardcoded string, so this isn't a regression — but for full I18n correctness, the "Leave this field blank..." portion could also be moved to a locale key.

Overall this is a clean, well-tested fix that aligns with existing Spree patterns. Nice work addressing the Copilot review feedback in the second commit.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by damianlegawiec — APPROVED** [review]
